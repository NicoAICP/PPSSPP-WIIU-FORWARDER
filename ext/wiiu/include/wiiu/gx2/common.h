#pragma once

#define GX2_SCAN_BUFFER_ALIGNMENT      0x1000
#define GX2_SHADER_ALIGNMENT           0x100
#define GX2_CONTEXT_STATE_ALIGNMENT    0x100
#define GX2_UNIFORM_BLOCK_ALIGNMENT    0x100
#define GX2_DISPLAY_LIST_ALIGNMENT     0x20
#define GX2_VERTEX_BUFFER_ALIGNMENT    0x40
#define GX2_INDEX_BUFFER_ALIGNMENT     0x20

#define GX2_ENABLE                     TRUE
#define GX2_DISABLE                    FALSE

#define GX2_TRUE                       TRUE
#define GX2_FALSE                      FALSE

#ifndef GX2_COMP_SEL
#define GX2_COMP_SEL(c0, c1, c2, c3) (((c0) << 24) | ((c1) << 16) | ((c2) << 8) | (c3))
#ifdef __cplusplus
#include <functional>
#endif
#include <stdio.h>

#define _x 0
#define _y 1
#define _z 2
#define _w 3
#define _r 0
#define _g 1
#define _b 2
#define _a 3
#define _0 4
#define _1 5
#endif


#define GX2_DISABLE_WRAPS
