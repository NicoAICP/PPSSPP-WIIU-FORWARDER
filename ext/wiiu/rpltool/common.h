#ifndef COMMON_H
#define COMMON_H

#include <ctype.h>
#include <stdint.h>
#include <stdio.h>

typedef int8_t s8;
typedef int16_t s16;
typedef int32_t s32;

typedef uint8_t u8;
typedef uint16_t u16;
typedef uint32_t u32;

#define GET_ENUM_NAME(prefix, name, val) \
   case prefix##name: \
      return #name;

#define EMIT_ENUM_TO_STR_FUNC(name, cases) \
   static inline const char *name##_to_str(name type) { \
      switch (type) { cases } \
      return "unknown"; \
   }

#define GET_FLAG_NAME(prefix, name, val) \
   case prefix##name: \
      pos += snprintf(buffer + pos, sizeof(buffer) - pos, "%s%s", #name, seperator); \
      flags &= ~(1u << i); \
      break;

#define EMIT_FLAGS_TO_STR_FUNC(name, cases) \
   static inline const char *name##_to_str(int flags, const char *seperator) { \
      int i; \
      static char buffer[32 * 32]; \
      int pos = 0; \
      for (i = 0; i < 32; i++) \
         if ((1u << i) & flags) \
            switch (1u << i) { cases } \
      if (flags) \
         snprintf(buffer + pos, sizeof(buffer) - pos, "0x%08X", flags); \
      else if (pos) \
         buffer[pos - strlen(seperator)] = '\0'; \
      else \
         return "NULL"; \
      return buffer; \
   }

#define MAKE_MAGIC(a, b, c, d) (((a) << 24) | ((b) << 16) | ((c) << 8) | ((d) << 0))

#define log_var(X) printf("%-35s: 0x%0*X (%u)\n", #X, sizeof(X) * 2, (u32)(X), (u32)(X))
#define log_svar(X) printf("%-35s: 0x%0*X (%i)\n", #X, sizeof(X) * 2, (u32)(X), (s32)(X))
#define log_array(X) debug_array(X, sizeof(X), #X)
#define log_enum(type, X) printf("%-35s: 0x%X (%s)\n", #X, (X), type##_to_str(X))
#define log_flags(type, X) printf("%-35s: 0x%08X (%s)\n", #X, (X), type##_to_str(X, " | "))
#define log_magic(X) \
   printf("%-35s: '%c''%c''%c''%c' (0x%08X)\n", #X, (u32)(X) >> 24, (u32)(X) >> 16, (u32)(X) >> 8, (u32)(X), (u32)(X))
#define log_carray(X) debug_char_array(X, sizeof(X), #X)
#define log_str(X) printf("%-35s: \"%s\"\n", #X, (char *)(X))
#define log_off_str(X, data) printf("%-35s: 0x%08X \"%s\"\n", #X, (X), (char *)data + (X))
#define log_off_strs(X, data) \
   do { \
      printf("%-35s: 0x%08X ", #X, (X)); \
      const char *ptr = (char *)data + (X); \
      while (*ptr) { \
         ptr += printf("\"%s\" ", ptr) - 2; \
      } \
      printf("\n"); \
   } while (0)

static inline void dump_array(uint32_t *src, int size);
static inline void debug_array(uint32_t *src, int size, const char *name) {
   printf("%-35s:", name);
   dump_array(src, size);
}
static inline void debug_char_array(uint8_t *src, int size, const char *name) {
   int i;
   printf("%-35s: ", name);
   //   for (i = 0; i < size; i++)
   //      printf("%-2c ", isalnum(src[i])?src[i]: '?');
   //   printf("\n%-35s: ", "");
   for (i = 0; i < size; i++)
      printf("%02X ", src[i]);
   printf("\n");
}

static inline void dump_array(uint32_t *src, int size) {
   int i;
   size = (size / 4) - 1;
   printf("\n{");
   for (i = 0; i < size; i++) {
      if (!(i & 3))
         printf("\n   ");

      printf("0x%08X, ", src[i]);
   }

   if (!(i & 3))
      printf("\n   ");

   printf("0x%08X\n}\n", src[i]);
}

static inline u32 align_up(u32 x, u32 align) { return (x + (align - 1)) & ~(align - 1); }
static inline u32 align_down(u32 x, u32 align) { return x & ~(align - 1); }

#endif // COMMON_H
