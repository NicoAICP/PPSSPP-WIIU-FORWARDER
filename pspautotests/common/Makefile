TARGET_LIB = libcommon.a
EXTRA_TARGETS = libcommon_kernel.a
OBJS = common.o vram.o sysmem-imports.o
KERNEL_OBJS = common_kernel.o vram.o sysmem-imports.o

INCDIR = 
CFLAGS = -O0 -G 0 -g -Wall
CXXFLAGS = $(CFLAGS) -fno-exceptions -fno-rtti
ASFLAGS = $(CFLAGS)

LIBDIR =
LDFLAGS = -G0

PSPSDK=$(shell psp-config --pspsdk-path)
include $(PSPSDK)/lib/build.mak

%.o: %.S
	$(AS) $(ASFLAGS) -c -o $@ $<

common_kernel.o: common.c
	$(CC) $(CFLAGS) -DCOMMON_KERNEL=1 -c -o $@ $<

libcommon_kernel.a: $(KERNEL_OBJS)
	$(AR) cru $@ $(KERNEL_OBJS)
	$(RANLIB) $@
