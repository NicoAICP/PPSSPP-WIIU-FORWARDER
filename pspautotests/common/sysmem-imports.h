extern int sceKernelSetCompiledSdkVersion(int ver);
extern int sceKernelSetCompiledSdkVersion370(int ver);
extern int sceKernelSetCompiledSdkVersion380_390(int ver);
extern int sceKernelSetCompiledSdkVersion395(int ver);
extern int sceKernelSetCompiledSdkVersion401_402(int ver);
extern int sceKernelSetCompiledSdkVersion500_505(int ver);
extern int sceKernelSetCompiledSdkVersion507(int ver);
extern int sceKernelSetCompiledSdkVersion600_602(int ver);
extern int sceKernelSetCompiledSdkVersion603_605(int ver);
extern int sceKernelSetCompiledSdkVersion606(int ver);
extern int sceKernelGetCompiledSdkVersion();