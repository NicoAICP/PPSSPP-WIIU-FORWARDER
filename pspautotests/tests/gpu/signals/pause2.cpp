#include <common.h>

#include <stdlib.h>
#include <stdio.h>
#include <pspgu.h>
#include <pspgum.h>
#include <pspthreadman.h>

extern "C" {
#include "sysmem-imports.h"
#include "../commands/commands.h"
}

static uint32_t __attribute__((aligned(16))) dlist1[] = {
	(GE_CMD_NOP << 24) | 0x000000,
	(GE_CMD_AMBIENTCOLOR << 24) | 0x000001,
	(GE_CMD_SIGNAL << 24) | (PSP_GE_SIGNAL_HANDLER_PAUSE << 16) | 0x0064,
	(GE_CMD_END << 24) | 0x000000,
	(GE_CMD_FINISH << 24) | 0x000000,
	(GE_CMD_END << 24) | 0x000000,
	(GE_CMD_AMBIENTCOLOR << 24) | 0x000002,
	(GE_CMD_FINISH << 24) | 0x000000,
	(GE_CMD_END << 24) | 0x000000,
	(GE_CMD_AMBIENTCOLOR << 24) | 0x000003,
	(GE_CMD_FINISH << 24) | 0x000000,
	(GE_CMD_END << 24) | 0x000000,
	(GE_CMD_AMBIENTCOLOR << 24) | 0x000004,
	(GE_CMD_FINISH << 24) | 0x000000,
	(GE_CMD_END << 24) | 0x000000,
};

static uint32_t __attribute__((aligned(16))) dlist2[] = {
	(GE_CMD_NOP << 24) | 0x000000,
	(GE_CMD_AMBIENTCOLOR << 24) | 0x000001,
	(GE_CMD_AMBIENTALPHA << 24) | 0x000001,
	(GE_CMD_SIGNAL << 24) | (PSP_GE_SIGNAL_HANDLER_PAUSE << 16) | 0x0064,
	(GE_CMD_END << 24) | 0x000000,
	(GE_CMD_AMBIENTCOLOR << 24) | 0x000002,
	(GE_CMD_FINISH << 24) | 0x000000,
	(GE_CMD_END << 24) | 0x000000,
	(GE_CMD_AMBIENTCOLOR << 24) | 0x000003,
	(GE_CMD_FINISH << 24) | 0x000000,
	(GE_CMD_END << 24) | 0x000000,
	(GE_CMD_AMBIENTCOLOR << 24) | 0x000004,
	(GE_CMD_FINISH << 24) | 0x000000,
	(GE_CMD_END << 24) | 0x000000,
};

static uint32_t __attribute__((aligned(16))) dlist3[] = {
	(GE_CMD_NOP << 24) | 0x000000,
	(GE_CMD_AMBIENTCOLOR << 24) | 0x000010,
	(GE_CMD_AMBIENTALPHA << 24) | 0x000010,
	(GE_CMD_FINISH << 24) | 0x000000,
	(GE_CMD_END << 24) | 0x000000,
};

static int currentListID = 0;
static uint32_t *currentListStart = NULL;

static int ge_signal(int value, void *arg, u32 *listpc) {
	if (listpc != NULL) {
		checkpoint("  * ge_signal: %08x +%08x", arg, listpc - currentListStart);
	} else {
		checkpoint("  * ge_signal: %08x %08x", arg, listpc);
	}
	checkpoint("  * ge_signal: %d at %x, listsync: %08x, drawsync: %08x", value, sceGeGetCmd(GE_CMD_AMBIENTCOLOR), sceGeListSync(currentListID, 1), sceGeDrawSync(1));
	return 0;
}

static int ge_signal2(int value, void *arg, u32 *listpc) {
	if (listpc != NULL) {
		checkpoint("  * ge_signal: %08x +%08x", arg, listpc - currentListStart);
	} else {
		checkpoint("  * ge_signal: %08x %08x", arg, listpc);
	}
	checkpoint("  * ge_signal: %d at %x, listsync: %08x, drawsync: %08x", value, sceGeGetCmd(GE_CMD_AMBIENTCOLOR), sceGeListSync(currentListID, 1), sceGeDrawSync(1));
	checkpoint("  * Continue: %08x", sceGeContinue());
	return 0;
}

static int ge_finish(int value, void *arg, u32 *listpc) {
	if (listpc != NULL) {
		checkpoint("  * ge_signal: %08x +%08x", arg, listpc - currentListStart);
	} else {
		checkpoint("  * ge_signal: %08x %08x", arg, listpc);
	}
	checkpoint("  * ge_finish: %d at %x, listsync: %08x, drawsync: %08x", value, sceGeGetCmd(GE_CMD_AMBIENTCOLOR), sceGeListSync(currentListID, 1), sceGeDrawSync(1));
	return 0;
}

static void testPauseSdkVersion(u32 ver) {
	char temp[256];
	snprintf(temp, sizeof(temp), "SDK version %08x", ver);
	checkpointNext(temp);
	if (ver != 0) {
		sceKernelSetCompiledSdkVersion(ver);
	}

	PspGeCallbackData cbdata;

	cbdata.signal_func = (PspGeCallback) ge_signal;
	cbdata.signal_arg  = NULL;
	cbdata.finish_func = (PspGeCallback) ge_finish;
	cbdata.finish_arg  = NULL;
	int cbid1 = sceGeSetCallback(&cbdata);

	currentListStart = dlist1;
	currentListID = sceGeListEnQueue(dlist1, dlist1, cbid1, NULL);
	sceGeListUpdateStallAddr(currentListID, dlist1 + 100);
	sceKernelDelayThread(10000);
	checkpoint("  UpdateStallAddr: %08x", sceGeListUpdateStallAddr(currentListID, dlist1 + 100));
	checkpoint("  Continue: %08x", sceGeContinue());
	sceKernelDelayThread(10000);
	sceGeBreak(1, NULL);
	checkpoint("  Callback 1 done: %x", sceGeGetCmd(GE_CMD_AMBIENTCOLOR));

	sceGeUnsetCallback(cbid1);

	cbdata.signal_func = (PspGeCallback) ge_signal2;
	cbdata.signal_arg  = NULL;
	cbdata.finish_func = (PspGeCallback) ge_finish;
	cbdata.finish_arg  = NULL;
	int cbid2 = sceGeSetCallback(&cbdata);

	currentListStart = dlist1;
	currentListID = sceGeListEnQueue(dlist1, dlist1, cbid2, NULL);
	sceGeListUpdateStallAddr(currentListID, dlist1 + 100);
	sceKernelDelayThread(10000);
	checkpoint("  UpdateStallAddr: %08x", sceGeListUpdateStallAddr(currentListID, dlist1 + 100));
	checkpoint("  Continue: %08x", sceGeContinue());
	sceKernelDelayThread(10000);
	sceGeBreak(1, NULL);
	checkpoint("  Callback 2 done: %x", sceGeGetCmd(GE_CMD_AMBIENTCOLOR));
	
	sceGeUnsetCallback(cbid2);
}

static void testPauseHead() {
	checkpointNext("Pause with queue head");

	PspGeCallbackData cbdata;

	cbdata.signal_func = (PspGeCallback)ge_signal;
	cbdata.signal_arg = NULL;
	cbdata.finish_func = (PspGeCallback)ge_finish;
	cbdata.finish_arg = NULL;
	int cbid1 = sceGeSetCallback(&cbdata);

	currentListStart = dlist2;
	currentListID = sceGeListEnQueue(dlist2, (uint8_t *)dlist2 + sizeof(dlist2), cbid1, NULL);
	sceKernelDelayThread(10000);
	checkpoint("  After pause: %x / %x", sceGeGetCmd(GE_CMD_AMBIENTCOLOR), sceGeGetCmd(GE_CMD_AMBIENTALPHA));

	int listid2 = sceGeListEnQueueHead(dlist3, (uint8_t *)dlist3 + sizeof(dlist3), -1, NULL);
	sceKernelDelayThread(10000);
	checkpoint("  After enqueue head: %x / %x", sceGeGetCmd(GE_CMD_AMBIENTCOLOR), sceGeGetCmd(GE_CMD_AMBIENTALPHA));

	checkpoint("  Continue: %08x", sceGeContinue());
	sceKernelDelayThread(10000);
	sceGeBreak(1, NULL);
	checkpoint("  Final list: %x / %x", sceGeGetCmd(GE_CMD_AMBIENTCOLOR), sceGeGetCmd(GE_CMD_AMBIENTALPHA));

	sceGeUnsetCallback(cbid1);
}

extern "C" int main(int argc, char *argv[]) {
	testPauseSdkVersion(0);
	testPauseSdkVersion(0x02080000);
	testPauseSdkVersion(0x03080000);
	testPauseSdkVersion(0x06060010);

	sceKernelSetCompiledSdkVersion(0x06060010);
	testPauseHead();

	return 0;
}