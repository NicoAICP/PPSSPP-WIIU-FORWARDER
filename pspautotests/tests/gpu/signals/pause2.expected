[x] SDK version 00000000
[x]   * ge_signal: 00000000 00000000
[x]   * ge_signal: 100 at 5c000001, listsync: 00000004, drawsync: 00000002
[x]   UpdateStallAddr: 00000000
[x]   * ge_signal: 00000000 00000000
[x]   * ge_finish: 0 at 5c000002, listsync: 00000000, drawsync: 00000000
[x]   Continue: 00000000
[r]   Callback 1 done: 5c000002
[x]   * ge_signal: 00000000 00000000
[x]   * ge_signal: 100 at 5c000001, listsync: 00000004, drawsync: 00000002
[x]   * Continue: 00000000
[x]   * ge_signal: 00000000 00000000
[x]   * ge_finish: 0 at 5c000002, listsync: 00000000, drawsync: 00000000
[x]   UpdateStallAddr: 80000020
[x]   Continue: 00000000
[r]   Callback 2 done: 5c000002

[x] SDK version 02080000
[x]   * ge_signal: 00000000 +00000006
[x]   * ge_signal: 100 at 5c000001, listsync: 00000004, drawsync: 00000002
[x]   UpdateStallAddr: 00000000
[x]   * ge_signal: 00000000 +00000009
[x]   * ge_finish: 0 at 5c000002, listsync: 00000000, drawsync: 00000000
[x]   Continue: 00000000
[r]   Callback 1 done: 5c000002
[x]   * ge_signal: 00000000 +00000006
[x]   * ge_signal: 100 at 5c000001, listsync: 00000004, drawsync: 00000002
[x]   * Continue: 00000000
[x]   * ge_signal: 00000000 +00000009
[x]   * ge_finish: 0 at 5c000002, listsync: 00000000, drawsync: 00000000
[x]   UpdateStallAddr: 80000020
[x]   Continue: 00000000
[r]   Callback 2 done: 5c000002

[x] SDK version 03080000
[x]   * ge_signal: 00000000 +00000006
[x]   * ge_signal: 100 at 5c000001, listsync: 00000004, drawsync: 00000002
[r]   UpdateStallAddr: 00000000
[x]   * ge_signal: 00000000 +00000009
[x]   * ge_finish: 0 at 5c000002, listsync: 00000000, drawsync: 00000000
[x]   Continue: 00000000
[r]   Callback 1 done: 5c000002
[x]   * ge_signal: 00000000 +00000006
[x]   * ge_signal: 100 at 5c000001, listsync: 00000004, drawsync: 00000002
[x]   * Continue: 00000000
[x]   * ge_signal: 00000000 +00000009
[x]   * ge_finish: 0 at 5c000002, listsync: 00000000, drawsync: 00000000
[r]   UpdateStallAddr: 80000020
[x]   Continue: 00000000
[r]   Callback 2 done: 5c000002

[x] SDK version 06060010
[x]   * ge_signal: 00000000 +00000006
[x]   * ge_signal: 100 at 5c000001, listsync: 00000004, drawsync: 00000002
[r]   UpdateStallAddr: 00000000
[x]   * ge_signal: 00000000 +00000009
[x]   * ge_finish: 0 at 5c000002, listsync: 00000000, drawsync: 00000000
[x]   Continue: 00000000
[r]   Callback 1 done: 5c000002
[x]   * ge_signal: 00000000 +00000006
[x]   * ge_signal: 100 at 5c000001, listsync: 00000004, drawsync: 00000002
[x]   * Continue: 00000000
[x]   * ge_signal: 00000000 +00000009
[x]   * ge_finish: 0 at 5c000002, listsync: 00000000, drawsync: 00000000
[r]   UpdateStallAddr: 80000020
[x]   Continue: 00000000
[r]   Callback 2 done: 5c000002

[x] Pause with queue head
[x]   * ge_signal: 00000000 +00000008
[x]   * ge_signal: 100 at 5c000002, listsync: 00000000, drawsync: 00000002
[r]   After pause: 5c000002 / 5d000001
[r]   After enqueue head: 5c000002 / 5d000001
[x]   * ge_signal: 00000000 +0000000b
[x]   * ge_finish: 0 at 5c000003, listsync: 00000000, drawsync: 00000000
[x]   Continue: 00000000
[r]   Final list: 5c000003 / 5d000010
